### PR TITLE
Request status block

### PR DESCRIPTION
**Pull request by Deren-Liao**
+ Showing a message loading or cancelling message.
+ A cancel button.  When doing advanced filtering, it can take a very long time.
+ Show errors , it is especially useful when the advanced filter syntax is wrong.

### CODE REVIEW

File: GoogleCloudExtension/GoogleCloudExtension/StackdriverLogsViewer/LogsViewerViewModel.cs [hunk 1]
```diff
@@ -147,6 +212,38 @@ public void LoadNextPage()
...
+            // Double check lock.
+            if (_isLoading)
+            {
+                lock (_isLoadingLockObj)
+                {
+                    if (_isLoading)
+                    {
+                        RequestStatusText = Resources.LogViewerRequestCancellingMessage;
+                        CancelRequestButtonVisibility = Visibility.Collapsed;
+                        _requestCancelled = true;
```
Comment (ivannaranjo) [inline GoogleCloudExtension/GoogleCloudExtension/StackdriverLogsViewer/LogsViewerViewModel.cs]: You are not handling cancellation the right way, you should be using `CancellationToken` and `CancellationTokenSource` to correctly cancel the requests. All of the Google APIs support this mechanism as well so you can actually cancel ongoing API calls, not only in between API calls.

File: GoogleCloudExtension/GoogleCloudExtension/StackdriverLogsViewer/LogsViewerViewModel.cs [hunk 2]
```diff
@@ -169,13 +266,40 @@ private void EnableControls()
         /// <param name="callback">A function to execute.</param>
         private async Task LogLoaddingWrapperAsync(Func<Task> callback)
         {
+            lock (_isLoadingLockObj)
```
Comment (ivannaranjo) [inline GoogleCloudExtension/GoogleCloudExtension/StackdriverLogsViewer/LogsViewerViewModel.cs]: Again, no need for locks.

File: GoogleCloudExtension/GoogleCloudExtension/StackdriverLogsViewer/LogsViewerViewModel.cs [hunk 3]
```diff
@@ -169,13 +266,40 @@ private void EnableControls()
...
             try
             {
-                DisableControls();
                 await callback();
             }
+            catch (DataSourceException ex)
+            {
+                RequestErrorMessage = ex.Message;
+            }
+            catch (Exception ex)
```
Comment (ivannaranjo) [inline GoogleCloudExtension/GoogleCloudExtension/StackdriverLogsViewer/LogsViewerViewModel.cs]: Bad idea, you should not handle `Exception`, only the ones that you can handle. Let the `ProtectedCommand` handle the rest, a dialog will be shown to the user.

The problem with handling `Exception` is that you are handling even the very bad ones, like `OutOfMemory`, which cannot be handled safely by our code.

File: GoogleCloudExtension/GoogleCloudExtension/StackdriverLogsViewer/LogsViewerViewModel.cs [hunk 4]
```diff
@@ -190,8 +314,9 @@ private void EnableControls()
         private async Task LoadLogsAsync()
         {
             int count = 0;
+            _requestCancelled = false;
```
Comment (ivannaranjo) [inline GoogleCloudExtension/GoogleCloudExtension/StackdriverLogsViewer/LogsViewerViewModel.cs]: You should be using a `CancellationTokenSource` here and pass the `CancelToken` around when performing calls.

File: GoogleCloudExtension/GoogleCloudExtension/StackdriverLogsViewer/LogsViewerToolWindowControl.xaml [hunk 5]
```diff
@@ -276,7 +276,37 @@
...
+            
+            <TextBox Margin="0, 0, 0, 12" 
+                     Padding="8, 5, 8, 5" 
+                     VerticalAlignment="Bottom" 
+                     TextAlignment="Left" 
+                     TextWrapping="Wrap"
+                     MaxHeight="120" 
+                     Foreground="Red" 
+                     Text="{Binding RequestErrorMessage, Mode=OneWay}" 
+                     Visibility="{Binding RequestErrorMessageVisibility}"
```
Comment (Deren-Liao) [inline GoogleCloudExtension/GoogleCloudExtension/StackdriverLogsViewer/LogsViewerToolWindowControl.xaml]: Done

File: GoogleCloudExtension/GoogleCloudExtension/StackdriverLogsViewer/LogsViewerToolWindowControl.xaml [hunk 6]
```diff
@@ -276,7 +276,37 @@
             </DataGrid>
         </Grid>
         
-        <!-- TODO: Displays log list requesting hint, or request error. -->
-        <Grid Grid.Row="3" Margin="12, 0, 12, 0" Background="White" />
+        <!-- Displays request status. -->
+        <Grid Grid.Row="3" Margin="12, 0, 12, 0" Background="White">
```
Comment (Deren-Liao) [inline GoogleCloudExtension/GoogleCloudExtension/StackdriverLogsViewer/LogsViewerToolWindowControl.xaml]: Done.

File: GoogleCloudExtension/GoogleCloudExtension/StackdriverLogsViewer/LogsViewerViewModel.cs [hunk 7]
```diff
@@ -33,6 +34,8 @@ public class LogsViewerViewModel : ViewModelBase
     {
         private const int DefaultPageSize = 100;
 
+        private object _isLoadingLockObj = new object();
```
Comment (Deren-Liao) [inline GoogleCloudExtension/GoogleCloudExtension/StackdriverLogsViewer/LogsViewerViewModel.cs]: Done

File: GoogleCloudExtension/GoogleCloudExtension/StackdriverLogsViewer/LogsViewerViewModel.cs [hunk 8]
```diff
@@ -92,6 +102,60 @@ public string FirstRowDate
...
 
         /// <summary>
+        /// Gets the cancel request button ICommand interface.
+        /// </summary>
+        public ProtectedCommand CancelRequestCommand { get; }
+
+        /// <summary>
+        /// Gets the cancel request button visibility
+        /// </summary>
+        public Visibility CancelRequestButtonVisibility
```
Comment (Deren-Liao) [inline GoogleCloudExtension/GoogleCloudExtension/StackdriverLogsViewer/LogsViewerViewModel.cs]: Done.

File: GoogleCloudExtension/GoogleCloudExtension/StackdriverLogsViewer/LogsViewerViewModel.cs [hunk 9]
```diff
@@ -92,6 +102,60 @@ public string FirstRowDate
...
+
+        /// <summary>
+        /// Gets the request error message.
+        /// </summary>
+        public string RequestErrorMessage
+        {
+            get { return _requestErrorMessage; }
+            private set {
+                SetValueAndRaise(ref _requestErrorMessage, value);
+                RequestErrorMessageVisibility = Visibility.Visible;
```
Comment (Deren-Liao) [inline GoogleCloudExtension/GoogleCloudExtension/StackdriverLogsViewer/LogsViewerViewModel.cs]: Done.

File: GoogleCloudExtension/GoogleCloudExtension/StackdriverLogsViewer/LogsViewerViewModel.cs [hunk 10]
```diff
@@ -147,6 +212,38 @@ public void LoadNextPage()
...
 
         /// <summary>
+        /// Cancel request button command.
+        /// </summary>
+        private void CancelRequest()
+        {
+            Debug.WriteLine("Cancel command is called");
+
+            // Double check lock.
+            if (_isLoading)
```
Comment (Deren-Liao) [inline GoogleCloudExtension/GoogleCloudExtension/StackdriverLogsViewer/LogsViewerViewModel.cs]: Done.

File: GoogleCloudExtension/GoogleCloudExtension/StackdriverLogsViewer/LogsViewerViewModel.cs (hunk 3 above)
Comment (Deren-Liao) [inline GoogleCloudExtension/GoogleCloudExtension/StackdriverLogsViewer/LogsViewerViewModel.cs]: Removed. I intended to have user open issue with the exception stack trace if it happens.

File: GoogleCloudExtension/GoogleCloudExtension/StackdriverLogsViewer/LogsViewerViewModel.cs (hunk 1 above)
Comment (Deren-Liao) [inline GoogleCloudExtension/GoogleCloudExtension/StackdriverLogsViewer/LogsViewerViewModel.cs]: Done

File: GoogleCloudExtension/GoogleCloudExtension/StackdriverLogsViewer/LogsViewerViewModel.cs (hunk 4 above)
Comment (Deren-Liao) [inline GoogleCloudExtension/GoogleCloudExtension/StackdriverLogsViewer/LogsViewerViewModel.cs]: Done.

**Comment by Deren-Liao** [comment]
I made changes as the feedback suggested.

**Pull request merged**